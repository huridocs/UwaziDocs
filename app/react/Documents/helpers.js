export default {
  prepareMetadata(doc, templates, thesauris) {

    let template = templates.find(t => t._id === doc.template);

    if (!template || !thesauris.length) {
      return Object.assign({}, doc, {metadata: [], documentType: ''});
    }

    let metadata = template.properties.map((property) => {
      let value = doc.metadata[property.name];
      if (property.type === 'select' && value) {
        value = thesauris.find(t => t._id === property.content).values.find(v => v.id === doc.metadata[property.name]).label;
      }
      return {label: property.label, value};
    });

    return Object.assign({}, doc, {metadata: metadata, documentType: template.name});
  }
};
