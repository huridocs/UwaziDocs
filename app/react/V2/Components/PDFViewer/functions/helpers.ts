const triggerScroll = (ref: React.RefObject<HTMLDivElement>, frameId: number): number => {
  let attempts = 0;
  let id = frameId;

  const attemptScroll = () => {
    if (attempts > 9) {
      return;
    }

    if (ref.current && ref.current.clientHeight > 0) {
      ref.current.scrollIntoView({ behavior: 'instant' });
      return;
    }

    attempts += 1;
    id = requestAnimationFrame(attemptScroll);
  };

  attemptScroll();
  return id;
};

export { triggerScroll };
