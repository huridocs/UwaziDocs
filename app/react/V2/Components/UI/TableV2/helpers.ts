import { UniqueIdentifier } from '@dnd-kit/core';
import { Row } from 'react-table';
import { cloneDeep } from 'lodash';
import { RowWithId, TableProps } from './Table';

const equalityById = (
  previousState: { id: UniqueIdentifier }[],
  newState: { id: UniqueIdentifier }[]
) => {
  let areEqual = true;

  if (previousState.length !== newState.length) {
    areEqual = false;
  } else {
    let idsSet = new Set(previousState.map(obj => obj.id));
    for (let obj of newState) {
      if (!idsSet.has(obj.id)) {
        areEqual = false;
        break;
      }
    }
  }

  return areEqual;
};

const getRowIds = <T extends RowWithId<T>>(data: TableProps<T>['dataState'][0]) => {
  const identifiers: { id: UniqueIdentifier; parentId?: string }[] = [];

  data.forEach(element => {
    identifiers.push({ id: element.rowId });

    if (element.subRows?.length) {
      element.subRows.forEach(subElement => {
        identifiers.push({ id: subElement.rowId, parentId: element.rowId });
      });
    }
  });

  return identifiers;
};

const dndSortHandler = <T extends RowWithId<T>>(
  currentState: TableProps<T>['dataState'][0],
  dataIds: { id: UniqueIdentifier; parentId?: string }[],
  activeId: string | number,
  overId: string | number
): TableProps<T>['dataState'][0] => {
  const state = cloneDeep(currentState);

  const activeParent = dataIds.find(dataId => dataId.id === activeId)?.parentId;
  const overParent = dataIds.find(dataId => dataId.id === overId)?.parentId;

  const activePosition = activeParent
    ? state
        .find(item => item.rowId === activeParent)
        ?.subRows?.findIndex(element => element.rowId === activeId)
    : state.findIndex(element => element.rowId === activeId);

  const droppedPosition = overParent
    ? state
        .find(item => item.rowId === overParent)
        ?.subRows?.findIndex(element => element.rowId === overId)
    : state.findIndex(element => element.rowId === overId);

  const element = activeParent
    ? state.find(item => item.rowId === activeParent)?.subRows?.splice(activePosition, 1)[0]
    : state.splice(activePosition, 1)[0];

  if (overParent) {
    state.find(item => item.rowId === overParent)?.subRows?.splice(droppedPosition, 0, element);
  } else {
    state.splice(droppedPosition, 0, element);
  }

  return state;
};

const sortHandler = <T extends RowWithId<T>>(rows: Row<T>[]) =>
  rows.map(row => {
    const { original, subRows } = row;
    if (subRows.length) {
      original.subRows = subRows.map(subRow => subRow.original);
    }
    return original;
  });

export { getRowIds, dndSortHandler, sortHandler, equalityById };
