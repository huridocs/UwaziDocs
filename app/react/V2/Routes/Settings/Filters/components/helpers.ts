import { ClientSettingsFilterSchema } from 'app/apiResponseTypes';
import { ClientTemplateSchema } from 'app/istore';

type LoaderData = {
  filters: ClientSettingsFilterSchema[] | undefined;
  templates: ClientTemplateSchema[];
};

const filterAvailableTemplates = (
  templates: ClientTemplateSchema[],
  filters?: ClientSettingsFilterSchema[]
) => {
  const usedTemplatesIds: string[] = [];

  filters?.forEach(filter => {
    if (filter.items) {
      filter.items.forEach(item => {
        usedTemplatesIds.push(item.id!);
      });
    }

    if (filter.id) {
      usedTemplatesIds.push(filter.id);
    }
  });

  return templates.filter(template => !usedTemplatesIds.includes(template._id));
};

const createNewFilters = (
  selectedTemplatesIds: string[],
  templates?: ClientTemplateSchema[]
): ClientSettingsFilterSchema[] => {
  const newFilters = selectedTemplatesIds.map(templateId => {
    const template = templates?.find(templ => templ._id === templateId);
    return { id: templateId, name: template?.name };
  });

  return newFilters;
};

const updateFilters = (
  newFilter: ClientSettingsFilterSchema,
  filters?: ClientSettingsFilterSchema[]
) => {
  let isNewFilter = true;

  const updatedFilters = filters?.map(filter => {
    if (filter.id === newFilter.id) {
      isNewFilter = false;
      return newFilter;
    }
    return filter;
  });

  if (isNewFilter) {
    return [...(updatedFilters || []), newFilter];
  }

  return updatedFilters;
};

const deleteFilters = (
  originalFilters?: ClientSettingsFilterSchema[],
  filtersToRemove?: (string | undefined)[]
) => {
  if (!filtersToRemove) {
    return originalFilters;
  }

  return originalFilters
    ?.map(filter => {
      if (filtersToRemove.includes(filter.id!)) {
        return {};
      }

      if (filter.items) {
        const nestedFilters = filter.items.filter(item => !filtersToRemove.includes(item.id!));
        return { ...filter, items: nestedFilters };
      }

      return { ...filter };
    })
    .filter(filter => {
      if (!filter.id) {
        return false;
      }
      if (filter.items && filter.items.length === 0) {
        return false;
      }
      return true;
    });
};

const sanitizeFilters = (filters?: ClientSettingsFilterSchema[]) => {
  const sanitizedFilters = filters?.map(filter => {
    const sanitizedFilter = { ...filter };

    if (filter.items) {
      sanitizedFilter.items = filter.items.map(
        (item: { id?: string; label?: string; _id?: string }) => {
          const sanitizedItem = { ...item };
          if (sanitizedItem._id) {
            delete sanitizedItem._id;
          }
          return sanitizedItem;
        }
      );
    }

    return sanitizedFilter;
  });

  return sanitizedFilters;
};

export type { LoaderData };
export {
  filterAvailableTemplates,
  createNewFilters,
  updateFilters,
  deleteFilters,
  sanitizeFilters,
};
