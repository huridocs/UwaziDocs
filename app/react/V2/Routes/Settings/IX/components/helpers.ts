import { ClientEntitySchema } from 'app/istore';
import { EntitySuggestionType } from 'shared/types/suggestionType';

// eslint-disable-next-line max-statements
const updateSuggestionsByEntity = (
  currentSuggestions: EntitySuggestionType[],
  updatedEntity?: ClientEntitySchema
): EntitySuggestionType[] => {
  if (!updatedEntity) {
    return currentSuggestions;
  }

  const suggestionToUpdate = currentSuggestions.find(
    currentSuggestion => currentSuggestion.entityId === updatedEntity._id
  );

  const propertyToUpdate = suggestionToUpdate?.propertyName;

  if (!suggestionToUpdate || !propertyToUpdate) {
    return currentSuggestions;
  }

  if (propertyToUpdate === 'title' && updatedEntity.title) {
    const newTitle = updatedEntity.title;
    suggestionToUpdate.currentValue = newTitle;
    suggestionToUpdate.entityTitle = newTitle;
    suggestionToUpdate.state.match = suggestionToUpdate.suggestedValue === newTitle;
  }

  if (
    propertyToUpdate !== 'title' &&
    updatedEntity.metadata &&
    updatedEntity.metadata[propertyToUpdate]?.length
  ) {
    const newValue = updatedEntity.metadata[propertyToUpdate]![0].value;
    suggestionToUpdate.currentValue = newValue;
    suggestionToUpdate.state.match = suggestionToUpdate.suggestedValue === newValue;
  }

  if (
    propertyToUpdate !== 'title' &&
    (!updatedEntity.metadata || !updatedEntity.metadata[propertyToUpdate]?.length)
  ) {
    suggestionToUpdate.currentValue = '';
    suggestionToUpdate.state.match = suggestionToUpdate.suggestedValue === '';
  }

  const suggestions = currentSuggestions.map(currentSuggestion => {
    if (currentSuggestion._id === suggestionToUpdate._id) {
      return suggestionToUpdate;
    }
    return currentSuggestion;
  });

  return suggestions;
};

const updateSuggestions = (
  currentSuggestions: EntitySuggestionType[],
  suggestionsToAccept: EntitySuggestionType[] | []
): EntitySuggestionType[] => {
  if (!suggestionsToAccept.length) {
    return currentSuggestions;
  }

  const acceptedSuggestions = suggestionsToAccept.map(suggestionToAccept => {
    const updated = { ...suggestionToAccept };
    updated.state.match = true;
    updated.currentValue = suggestionToAccept.suggestedValue;

    if (
      suggestionToAccept.propertyName === 'title' &&
      typeof suggestionToAccept.suggestedValue === 'string'
    ) {
      updated.entityTitle = suggestionToAccept.suggestedValue;
    }

    return updated;
  });

  const merged = [
    ...currentSuggestions
      .concat(acceptedSuggestions)
      .reduce(
        (map, suggestion) =>
          map.set(suggestion._id, Object.assign(map.get(suggestion._id) || {}, suggestion)),
        new Map()
      )
      .values(),
  ];

  return merged;
};

export { updateSuggestions, updateSuggestionsByEntity };
