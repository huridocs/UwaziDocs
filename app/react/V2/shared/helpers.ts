import moment from 'moment-timezone';

const propertyValueFormatter = {
  date: (timestamp: string) => moment.utc(timestamp, 'X').format('ll'),
};

export { propertyValueFormatter };
