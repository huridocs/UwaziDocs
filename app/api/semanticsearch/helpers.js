import { entityDefaultDocument } from 'shared/entityDefaultDocument';
import { files } from 'api/files';

import { search } from '../search';
import model from './model';
import resultsModel from './resultsModel';
import { COMPLETED } from './statuses';
import templates from '../templates';

export const getSearchDocuments = async ({ documents, query }, language, user) => {
  if (documents && documents.length) {
    return documents;
  }
  const _query = { ...query, limit: 9999, searchTerm: '' };
  const res = await search.search(_query, language, user);
  return res.rows.map(doc => doc.sharedId);
};

export const removePageAnnotations = text => text.replace(/\[\[\d+\]\]/g, '');

export const updateSearchDocumentStatus = async (searchId, sharedId, status) =>
  model.db.findOneAndUpdate(
    {
      _id: searchId,
      'documents.sharedId': sharedId,
    },
    {
      $set: { 'documents.$.status': status },
    },
    { new: true, lean: true }
  );

export const setSearchDocumentResults = async (searchId, sharedId, results) => {
  const averageScore = results.length
    ? results.reduce((total, curr) => total + curr.score, 0) / results.length
    : 0;
  const _results = [...results];
  const docResults = await resultsModel.db.findOneAndUpdate(
    {
      sharedId,
      searchId,
    },
    {
      sharedId,
      searchId,
      averageScore,
      results: _results.sort((r1, r2) => r2.score - r1.score),
      status: COMPLETED,
    },
    { upsert: true, new: true }
  );
  return docResults;
};

export const extractDocumentContent = async (
  { documents, language, template, metadata },
  defaultLanguage
) => {
  const contents = {};
  let fullText;
  const defaultDocument = entityDefaultDocument(documents, language, defaultLanguage);
  if (defaultDocument && defaultDocument._id) {
    [{ fullText }] = await files.get(defaultDocument, '+fullText');
  }

  if (metadata) {
    (await templates.getById(template)).properties
      .filter(prop => prop.type === 'markdown')
      .forEach(field => {
        if (metadata[field.name]) {
          contents[field.name] = metadata[field.name];
        }
      });
  }

  if (fullText) {
    Object.keys(fullText).forEach(page => {
      contents[page] = removePageAnnotations(fullText[page]);
    });
  }

  return contents;
};
