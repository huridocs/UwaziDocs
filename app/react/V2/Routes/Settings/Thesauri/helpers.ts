import { ThesaurusSchema, ThesaurusValueSchema } from 'shared/types/thesaurusType';
import { httpRequest } from 'shared/superagent';
import { ClientThesaurusValue } from 'app/apiResponseTypes';

const sanitizeThesaurusValues = (
  thesaurus: ThesaurusSchema,
  values: ClientThesaurusValue[]
): ThesaurusSchema => {
  const sanitizedThesaurus = { ...thesaurus, values };
  // @ts-ignore
  sanitizedThesaurus.values =
    values?.map(sValue => {
      delete sValue.id;
      // @ts-ignore
      delete sValue.groupId;
      if (sValue.values) {
        sValue.values = sValue.values.map((ssValue: any) => {
          delete ssValue._id;
          // @ts-ignore
          delete ssValue.groupId;
          return ssValue;
        });
      }
      return sValue;
    }) || ([] as ThesaurusValueSchema[]);
  return sanitizedThesaurus;
};

function sanitizeThesauri(thesaurus: ThesaurusSchema) {
  const sanitizedThesauri = { ...thesaurus };
  sanitizedThesauri.values = sanitizedThesauri
    .values!.filter((value: ThesaurusValueSchema) => value.label)
    .filter((value: ThesaurusValueSchema) => !value.values || value.values.length)
    .map((value: ThesaurusValueSchema) => {
      const _value = { ...value };
      if (_value.values) {
        _value.values = _value.values.filter(_v => _v.label);
      }
      return _value;
    });
  return sanitizedThesauri;
}

const importThesaurus = async (
  thesaurus: ThesaurusSchema,
  file: File
): Promise<ThesaurusSchema> => {
  const headers = {
    Accept: 'application/json',
    'X-Requested-With': 'XMLHttpRequest',
  };
  const fields = {
    thesauri: JSON.stringify(thesaurus),
  };

  return (await httpRequest('thesauris', fields, headers, file)) as ThesaurusSchema;
};

export { sanitizeThesaurusValues, sanitizeThesauri, importThesaurus };
