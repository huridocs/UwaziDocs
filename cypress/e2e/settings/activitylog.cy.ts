import 'cypress-axe';
import { clearCookiesAndLogin } from '../helpers';

describe('Activity log', () => {
  // eslint-disable-next-line max-statements
  before(() => {
    const env = { DATABASE_NAME: 'uwazi_e2e', INDEX_NAME: 'uwazi_e2e' };
    cy.exec('yarn e2e-fixtures', { env });
    clearCookiesAndLogin('editor', 'editor');
    cy.get('.only-desktop a[aria-label="Settings"]').click();
    cy.contains('span', 'Account').click();
    cy.contains('button', 'Enable').click();
    cy.contains('button', 'Cancel').click();
    cy.clearAndType('input[name=email]', 'editor@uwazi.com');
    cy.contains('button', 'Update').click();

    clearCookiesAndLogin();
    cy.get('.only-desktop a[aria-label="Settings"]').click();
    cy.contains('span', 'Activity log').click();
    cy.contains('editor');
    cy.injectAxe();
  });

  it('should pass accessibility check', () => {
    cy.checkA11y();
  });

  const checkCells = (row: number, column: number, element: string) => {
    cy.get(`tr:nth-child(${row}) > td:nth-child(${column}) > ${element}`).toMatchSnapshot();
  };

  it('should list the last activity log entries', () => {
    cy.get('tr').should('have.length.at.least', 10);
    checkCells(1, 1, 'span');
    checkCells(1, 2, 'span');
    checkCells(1, 3, ' div > div:nth-child(1)');
    checkCells(2, 1, 'span');
    checkCells(2, 2, 'span');
    checkCells(2, 3, ' div > div:nth-child(1)');
  });

  it('should filter by user', () => {
    cy.clearAndType('input[name=username]', 'editor');
    cy.contains('Updated user');
    cy.get('tr').should('have.length.at.most', 5);
  });

  it('should show a tooltip with the detail of an activity entry', () => {
    cy.get('input[name=username]').clear();
    cy.contains('admin');
    cy.get('tr:nth-child(1) td:nth-child(3)').within(() => {
      cy.contains('Updated user').invoke('show').trigger('mouseenter');
    });
    cy.get('tr:nth-child(1) td:nth-child(3) [data-testid=flowbite-tooltip]')
      .should('be.visible')
      .invoke('text')
      .should('match', /Query{}Body{.+"username":"editor".+"email":"editor@uwazi.com"}/);
    cy.get('tr:nth-child(1) td:nth-child(3)').within(() => {
      cy.contains('Updated user').trigger('mouseleave');
    });
  });

  it('should update the list when filters are cleaned', () => {
    cy.get('tr').should('have.length.at.least', 10);
  });

  it('should open the detail of an entry', () => {
    cy.get('input[name=username]').clear();
    cy.get('tr:nth-child(1) td:nth-child(5)').contains('View').click();
    cy.get('aside').eq(0).toMatchImageSnapshot();
    cy.get('[data-testid=close-sidepanel]').click();
  });

  it('should filter by method', () => {
    cy.clearAndType('input[name=search]', 'DELETE');
    cy.contains('editor');
    cy.get('tr').should('have.length.at.most', 3);
  });

  it('should filter by dates', () => {
    cy.get('input[name=search]').clear();
    cy.get('input[name=from]').type('2023-06-22', { delay: 0 });
    cy.get('input[name=to]').clear();
    cy.get('input[name=to]').type('2023-06-23', { delay: 0 });
    cy.contains('admin');
    cy.get('tr').should('have.length.at.most', 2);
  });
});
