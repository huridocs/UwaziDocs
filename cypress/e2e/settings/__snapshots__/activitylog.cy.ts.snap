exports[`Activity log > should list the last activity log entries #0`] = `
<span
  class=" bg-blue-100 px-2.5 py-1 rounded-md text-xs bg-blue-100 text-blue-800"
  data-testid="pill-comp"
  ><span class="translation">UPDATE</span></span
>
`;

exports[`Activity log > should list the last activity log entries #1`] = `
<span class="text-primary-700">editor</span>
`;

exports[`Activity log > should list the last activity log entries #2`] = `
<div class="w-fit" data-testid="flowbite-tooltip-target">
  <div class="space-x-2">
    <span class="translation font-semibold">Updated user</span>:<span
      class="translation"
      >editor (58ada34d299e82674854510e)</span
    >
  </div>
</div>
`;

exports[`Activity log > should list the last activity log entries #3`] = `
<span
  class=" bg-blue-100 px-2.5 py-1 rounded-md text-xs bg-green-100 text-green-800"
  data-testid="pill-comp"
  ><span class="translation">CREATE</span></span
>
`;

exports[`Activity log > should list the last activity log entries #4`] = `
<span class="text-primary-700">editor</span>
`;

exports[`Activity log > should list the last activity log entries #5`] = `
<div class="w-fit" data-testid="flowbite-tooltip-target">
  <div class="space-x-2">
    <span class="translation font-semibold"
      >Two-factor authentication secret</span
    >:
  </div>
</div>
`;
